### PR TITLE
AMP Bootstrap

### LINKED ISSUE #306: Need an amp-bootstrap agent
To replace appcelerator/amp-swarm-deploy, we need a go binary that will take care of same operations:
- make sure the Docker engine is up to date
- initialize the Swarm cluster

Later if will be also used to register on the AMP portal

### PR DESCRIPTION
**Pull request by ndegory**
resolves #306 

subcommand of amp cluster cli to bootstrap a node (Docker, Swarm, cluster agent and server services)

```
$ amp-cluster bootstrap -h
Installs Docker, setups Swarm, run the AMP agent

Usage:
  amp-cluster bootstrap [flags]

Aliases:
  bootstrap, bs


Flags:
  -c, --create         initializes the Swarm (needs option -m)
      --host string    Swarm manager hostname
  -m, --manager        bootstraps a Swarm manager node
      --port string    Swarm manager port (default "2377")
      --tag string     AMP image version (default "v0.4.0-dev")
      --token string   Swarm join token

Global Flags:
      --config string   Config file (default is $HOME/.amp.yaml)
      --debug           display debug information
      --server string   define the remote server address (default "127.0.0.1:31315")
      --silence         No output
  -v, --verbose         Verbose output
```

### CODE REVIEW

**Review by generalhenry — APPROVED** [review]

**Comment by ndegory** [comment]
tested: manager initialization (sudo amp-cluster bootstrap -m -c --tag cluster) on Ubuntu 16.04 and CentOS 7.

File: cmd/amp-cluster/bootstrap.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	},
+}
+
+// Flag variables
+var SwarmManagerHost string
+var SwarmManagerPort string
+var SwarmJoinToken string
+var AmpImageTag string
+var CreateSwarm bool
+var Manager bool
```
Comment (freignat91) [inline cmd/amp-cluster/bootstrap.go]: These vars should have their first letter lowercase concidering they are only local

File: cmd/amp-cluster/bootstrap.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+			log.Fatalf("Failed to validate Docker version: %s\n", err)
+		}
+		if !ok {
+			log.Fatalf("Docker version does not meet the requirements: %s\n", version)
+		}
+	} else {
+		log.Println("Docker engine version is just fine")
+	}
+	if err := c.enableDockerEngine(); err != nil {
+		log.Fatalf("Failed to enable the Docker service: %s\n", err)
```
Comment (freignat91) [inline cmd/amp-cluster/bootstrap.go]: use g.fatalc("%v\n", err) in place of log.Fatalln to have color output
globally use g.printfc(...) in place of log to have color output the same way then in the orther commands
including in bootstrap-docker.go file

File: cmd/amp-cluster/bootstrap-docker.go [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+			return version.Version, nil
+		}
+	}
+	log.Println("Unable to contact the Docker server through the API: %s, switch to the Docker cli")
+	var out bytes.Buffer
+	var stderr bytes.Buffer
+	versionCmd := [...]string{"docker", "version", "-f", "{{ .Server.Version }}"}
+	cmd := exec.Command(versionCmd[0], versionCmd[1:]...)
+	cmd.Stdout = &out
+	cmd.Stderr = &stderr
```
Comment (freignat91) [inline cmd/amp-cluster/bootstrap-docker.go:53]: if you want to see output you can also use:
    cmd.Stdout = os.Stdout
    cmd.Stderr = os.Stderr

File: cmd/amp-cluster/bootstrap-docker.go [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+	_, err = c.dockerClient.SwarmInit(c.ctx, req)
+	if err == nil {
+		log.Println("done")
+	}
+	return err
+}
+
+func (c *clusterClient) pullAmpImage(version string) error {
+	image := "appcelerator/amp:" + version
+	_, err := c.dockerClient.ImagePull(c.ctx, image, types.ImagePullOptions{})
```
Comment (freignat91) [inline cmd/amp-cluster/bootstrap-docker.go]: you don't use the reader, so you don't know if pull has error or not during pull

**Review by freignat91 — CHANGES_REQUESTED** [review]

File: cmd/amp-cluster/bootstrap.go (hunk 1 above)
Comment (ndegory) [reply cmd/amp-cluster/bootstrap.go]: ok, done

File: cmd/amp-cluster/bootstrap.go (hunk 2 above)
Comment (ndegory) [reply cmd/amp-cluster/bootstrap.go]: ok, done (but now I have to background color issue...)

File: cmd/amp-cluster/bootstrap-docker.go (hunk 3 above)
Comment (ndegory) [reply cmd/amp-cluster/bootstrap-docker.go:53]: I can't, I don't want to output it on the console, I use the output later

File: cmd/amp-cluster/bootstrap-docker.go (hunk 4 above)
Comment (ndegory) [reply cmd/amp-cluster/bootstrap-docker.go]: ok, done

**Review by freignat91 — APPROVED** [review]
LGTM

**Pull request merged**